sealir-tutorials/ch08_gpu_offload.py

```diff
@@ -0,0 +1,158 @@
+        pass_man.add("scf-parallel-loop-fusion")
+        pass_man.add('func.func(gpu-map-parallel-loops,convert-parallel-loops-to-gpu)')
+        pass_man.add("gpu-kernel-outlining")
+        pass_man.add('gpu-lower-to-nvvm-pipeline{cubin-format="fatbin"}')
+        pass_man.add("convert-scf-to-cf")
+        pass_man.add("finalize-memref-to-llvm")
+        pass_man.add("convert-func-to-llvm")
+        pass_man.add("convert-index-to-llvm")
+        pass_man.add("convert-bufferization-to-memref")
+        pass_man.add("reconcile-unrealized-casts")
+        pass_man.add("func.func(llvm-request-c-wrappers)")
+        pass_man.enable_verifier(True)
+        pass_man.run(module.operation)
+        # Output LLVM-dialect MLIR
+        module.dump()
+        return module
+
+    @classmethod
+    def get_exec_ptr(cls, mlir_ty, val):
+        if isinstance(mlir_ty, ir.IntegerType):
+            val = 0 if val is None else val
+            ptr = ctypes.pointer(ctypes.c_int64(val))
+        elif isinstance(mlir_ty, ir.F32Type):
+            val = 0.0 if val is None else val
+            ptr = ctypes.pointer(ctypes.c_float(val))
+        elif isinstance(mlir_ty, ir.F64Type):
+            val = 0.0 if val is None else val
+            ptr = ctypes.pointer(ctypes.c_double(val))
+        elif isinstance(mlir_ty, ir.MemRefType):
+            val = np.zeros(mlir_ty.shape, dtype=np.float64) if val is None else val
+            val = cls.np_arr_to_np_duck_device_arr(val)
+            ptr = ctypes.pointer(ctypes.pointer(runtime.get_ranked_memref_descriptor(val)))
+
+        return ptr, val
+    
+    @classmethod
+    def get_out_val(cls, res_ptr, res_val):
+        if isinstance(res_val, cuda.cudadrv.devicearray.DeviceNDArray):
+            return res_val.copy_to_host()
+        else:
+            return super().get_out_val(res_ptr, res_val)
+
+    @classmethod
+    def np_arr_to_np_duck_device_arr(cls, arr):
+        da = cuda.to_device(arr)
+        ctlie = namedtuple("ctypes_lie", "data data_as shape")
+        da.ctypes = ctlie(da.__cuda_array_interface__["data"][0],
+                    lambda x: ctypes.cast(da.ctypes.data, x),
+                    da.__cuda_array_interface__["shape"],)
+        da.itemsize = arr.itemsize
+        return da
+
+    @classmethod
+    def jit_compile_(cls, llmod, input_types, output_types, function_name="func", exec_engine=None, **execution_engine_params):
+        cuda_libs = ("mlir_cuda_runtime", "mlir_c_runner_utils", "mlir_runner_utils")
+        cuda_shared_libs = [find_library(x) for x in cuda_libs]
+        return super().jit_compile_(llmod, input_types, output_types, function_name, exec_engine=execution_engine.ExecutionEngine(llmod, opt_level=3, shared_libs=cuda_shared_libs))
+
+
+compiler.set_backend(GPUBackend())
+compiler.set_converter_class(ConditionalExtendGraphtoRVSDG)
+compiler.set_cost_model(MyCostModel())
+
+
+@ufunc_vectorize(input_types=[Float64, Float64, Float64], shape=(10, 10))
```

Comment:
What if I want to switch between GPU and CPU backend in a single script? Couldn't compiler be passed into `ufunc_vectorize`?